Redis v2 changes

craigyam opened the pull request:
Switch to key/value strings to use automatic expiration.  Update the key to include keywords.

craigyam committed:
Registry Redis support v2.

Modify schema to key/value strings instead of hash to allow
use of auto-expiration.  Remove registry expiration code and tests.

craigyam committed:
Update redis record key to include keywords.

Add keywords to redis record in the following format:
reg:inst:<instance id>:ns:<namespace>

zcahana commented on registry/store/external_catalog_test.go:
```diff
@@ -428,55 +427,6 @@ func TestExternalListServices(t *testing.T) {
 	}
 }
 
-func TestExternalOutOfServiceDoesNotExpire(t *testing.T) {
```
Most of the deleted test cases are still valid. I guess they were removed because the redis mock doesn't support expirations. How complex would it be to add that support?

zcahana commented on registry/store/redis_registry.go:
```diff
@@ -42,50 +42,56 @@ func NewRedisRegistry(db database.Database) ExternalRegistry {
-	if len(keys) == 0 {
+	if len(entry) == 0 {
 		var emptyServiceInstance ServiceInstance
 		return &emptyServiceInstance, nil
```
Is it possible that `err == nil` but `len(entry) == 0`?
Either way, we should return an error in this case, and not an `emptyServiceInstance`.

craigyam replied:
len(entry) == 0 is a valid response (registering a new instance for example).  I returned an empty instance because we check the instance.ID during registration.  I can instead return nil and check instance != nil.

craigyam replied:
Done

zcahana commented on registry/store/external_registry.go:
```diff
@@ -24,18 +24,18 @@ import (
 	ListAllServiceInstances(namespace auth.Namespace) (map[string]ServiceInstanceMap, error)
 	InsertServiceInstance(namespace auth.Namespace, instance *ServiceInstance) error
-	DeleteServiceInstance(namespace auth.Namespace, key string) (int, error)
+	DeleteServiceInstance(key string) (int, error)
```
Should the `DeleteServiceInstance()` method accept a namespace param, as all other methods in this interface, instead of expecting the client (`externalCatalog`) to encode it?
Also, what's the significance of the `int` return value? Seems to be the numeric return value of redis... which doesn't make sense as part of this interface.

zcahana replied:
Sorry, forgot to review the 2nd commit, which already corrected the namespace parameter.

craigyam replied:
The return value is the number of entries deleted (0 is a valid value).  The calling code checks if the delete returned 0 and will respond that there was no service to delete.

zcahana commented on registry/store/redis_registry.go:
```diff
@@ -42,50 +42,56 @@ func NewRedisRegistry(db database.Database) ExternalRegistry {
 	}
 
-	return si, nil
+	return si.DeepClone(), nil
```
No reason to `DeepClone()` here. We've been using it for defensive copying when returning stored service instances, but in here, `si` is just a local variable that otherwise would be garbage collected.

craigyam replied:
Done

zcahana commented on registry/store/redis_registry.go:
```diff
@@ -42,50 +42,56 @@ func NewRedisRegistry(db database.Database) ExternalRegistry {
 			// Log an error, but continue with empty SI
 			rr.logger.WithFields(log.Fields{
 				"key": key,
 			}).Error("Unable to unmarshal json")
```
Should `continue` in case of an unmarshaling error.

craigyam replied:
Done

zcahana commented on registry/store/redis_registry.go:
```diff
@@ -112,11 +118,11 @@ func (rr *redisRegistry) ListAllServiceInstances(namespace auth.Namespace) (map[
 		}
 		siMap[siKey] = &si
-		serviceMap[siKey.ServiceName] = siMap
+		serviceMap[si.ServiceName] = siMap
```
Seems to be a bug here: `serviceMap` might already contain a `ServiceInstanceMap` entry for that key (service name). In that case, it will be overwritten. The end result is that for each service, only one instance will be returned.

craigyam replied:
Done

zcahana commented on registry/store/redis_registry.go:
```diff
@@ -125,11 +131,18 @@ func (rr *redisRegistry) ListAllServiceInstances(namespace auth.Namespace) (map[
-	siKey := fmt.Sprintf("%s.%s", instance.ID, instance.ServiceName)
+	siKey := fmt.Sprintf("%s%s%s", instance.ID, keySeparator, namespace.String())
+
+	err := rr.db.InsertEntry(siKey, instanceJSON)
```
In case of an error in `InsertEntry()`, we shouldn't `Expire()`.

craigyam replied:
Done

zcahana commented on registry/store/redis_registry.go:
```diff
@@ -125,11 +131,18 @@ func (rr *redisRegistry) ListAllServiceInstances(namespace auth.Namespace) (map[
+
+	// If the status is OUT_OF_SERVICE do not expire
+	if instance.Status != OutOfService {
+		rr.db.Expire(siKey, instance.TTL)
```
`Expire()` might fail as well. By minimum, this error should be logged.

craigyam replied:
Done

zcahana commented on registry/utils/database/redis_database.go:
```diff
@@ -125,80 +130,88 @@ func (rdb *redisDB) ReadAllEntries(hashname string) (map[string]string, error) {
+
+	matches, err := rdb.scan(conn, match)
+
+	for _, value := range matches {
```
Calling `GET` for each instance is extremely inefficient. [`MGET`](http://redis.io/commands/mget) is a much better option.

craigyam replied:
Done

zcahana reviewed (changes requested):
@craigyam Thanks for submitting the PR. Please see attached comments.

zcahana commented on registry/store/external_registry.go:
```diff
@@ -27,15 +27,24 @@ type ExternalRegistry interface {
+// reg:inst:<instance id>:ns:<namespace>
 func (dbk *DBKey) String() string {
-	return fmt.Sprintf("%s%s%s", dbk.InstanceID, keySeparator, dbk.Namespace)
+	return fmt.Sprintf("%s%s%s%s%s%s%s%s%s", registryRecordKey, keySeparator, instanceKey, keySeparator, dbk.InstanceID, keySeparator, namespaceKey, keySeparator, dbk.Namespace)
```
Alternative implementation that avoid the error-prone `%s%s%s%s%s%s%s%s%s`:

``` go
parts := []string{registryRecordKey, instanceKey, dbk.InstanceID, namespaceKey, dbk.Namespace}
return strings.Join(parts, keySeparator)
```

craigyam replied:
So I changed the DB key to be a JSON string which makes parsing the key read from the db way easier (just an Unmarshal vs. splitting the string based on the separators - the concern being handling separators that may appear in the values).  This should make adding/removing/moving fields in the key trivial compared to maintaining a parsing function.  Let me know what you think of the change.  If you prefer it the other way, I can change it back.

zcahana replied:
Assuming we'll have to support existing keys - adding/removing/moving fields in the key will be impossible, and at minimum not trivial at all. That also means that we must have control over the JSON marshaling - whitespace, fields ordering. `json.Marshal()` doesn't specify that, so we'll have to implement `MarshalJSON()` ourselves, including escaping `"` characters from the values etc. That also means larger keys (JSON overhead) and extra work to encode. The only benefit is we won't be needing to implement decoding on our own - the off-the-shelf `json.Unmarshal()` will do fine - but given the other issues, I suggest we stick to a simple (non-JSON) encoding scheme, and escape the `:` separators if present.

zcahana replied:
On another note, need to take into account that when doing SCAN, the namespace might contain glob-characters as well (`*`, `?`, ...). These must be escaped as well.

craigyam replied:
Switched back to the original key format and added a parser.  Also added a function to escape the glob chars.

zcahana commented on registry/store/external_registry.go:
```diff
@@ -27,15 +27,24 @@ type ExternalRegistry interface {
+// reg:inst:<instance id>:ns:<namespace>
 func (dbk *DBKey) String() string {
-	return fmt.Sprintf("%s%s%s", dbk.InstanceID, keySeparator, dbk.Namespace)
+	return fmt.Sprintf("%s%s%s%s%s%s%s%s%s", registryRecordKey, keySeparator, instanceKey, keySeparator, dbk.InstanceID, keySeparator, namespaceKey, keySeparator, dbk.Namespace)
```
Also, given that there's no good reason otherwise, I would suggest to order the key elements as follows:
`reg:ns:<namespace>:inst:<instance id>`.

This should already improve the performance of regex matching during scans, and could potentially improve even fruther in the future, in case redis will support optimized scans over sorted key space.

Also, it feels more natural and hierarchical this way.

craigyam replied:
See above comment.

craigyam replied:
Done

craigyam committed:
Code review comments for redis v2 changes.

Made changes based on code review comments.  Changed the key to be
a JSON string. Used MGET instead of GET for ReadAllEntries().

craigyam committed:
Merge branch 'master' into redis_v2

craigyam committed:
Change key back to string and add unit tests.

Change the db key back to a string from JSON and add a parser.
Also add some additional unit tests.

zcahana commented:
@craigyam Looks good, thanks!
